cpp/src/arrow/compute/kernels/scalar_temporal_test.cc

```diff
@@ -216,5 +216,40 @@ TEST_F(ScalarTemporalTest, DayOfWeek) {
                 DayOfWeek(timestamps, DayOfWeekOptions(/*one_based_numbering=*/false,
                                                        /*week_start=*/8)));
 }
+
+#ifndef _WIN32
+TEST_F(ScalarTemporalTest, Strftime) {
+  auto options_seconds = StrftimeOptions("%Y-%m-%dT%H:%M:%S%z", "UTC");
+  auto options_milliseconds = StrftimeOptions("%Y-%m-%dT%H:%M:%S%z", "GMT");
+  auto options_microseconds = StrftimeOptions("%Y-%m-%dT%H:%M:%S%z", "Asia/Kolkata");
+  auto options_nanoseconds = StrftimeOptions("%Y-%m-%dT%H:%M:%S%z", "US/Hawaii");
+
+  const char* times_seconds = R"(["1970-01-01T00:00:59", null])";
+  const char* times_milliseconds = R"(["1970-01-01T00:00:59.123", null])";
+  const char* times_microseconds = R"(["1970-01-01T00:00:59.123456", null])";
+  const char* times_nanoseconds = R"(["1970-01-01T00:00:59.123456789", null])";
+
+  const char* zoned_seconds = R"(["1970-01-01T00:00:59+0000", null])";
+  const char* zoned_milliseconds = R"(["1970-01-01T00:00:59.123+0000", null])";
```

Comment:
With a default format of `%S`, I don't expect sub-second values to be shown (since there is a `%f` for this) ?

Reply:
I see you commented about this in the python/pandas tests below as well. 
Is this the default behaviour of `date.h` (which seems a strange deviation from the standard?)

Reply:
`date.h` doesn't appear to have a `%f` flag.

[date.h](https://howardhinnant.github.io/date/date.html) doc states:

> ***%S*** | Seconds as a decimal number. If the number of seconds is less than 10, the result is prefixed with 0. If the precision of the input can not be exactly represented with seconds, then the format is a decimal floating point number with a fixed format and a precision matching that of the precision of the input (or to a microseconds precision if the conversion to floating point decimal seconds can not be made within 18 fractional digits). The character for the decimal point is localized according to the locale. The modified command %OS produces the locale's alternative representation.

Which almost exactly matches [c++ formatter](https://en.cppreference.com/w/cpp/chrono/duration/formatter).

Implementing [python-like](https://strftime.org/) `%f` and `%S` would require additional logic here.

R appears to have a [similar convention](https://stat.ethz.ch/R-manual/R-devel/library/base/html/strptime.html) to Python.

[SQLite](https://sqlite.org/lang_datefunc.html):
```
%f |   | fractional seconds: SS.SSS
%S |   | seconds: 00-59
```
[Postgres](https://www.postgresql.org/docs/9.1/ecpg-pgtypes.html):
```
%S - is replaced by the second as a decimal number (00-60).
No fractions.
```
[Ruby](https://apidock.com/ruby/DateTime/strftime):
```
%S - Second of the minute (00..59)
No fractions.
```

I kind of like the current behavior but it could would of course not cover all cases.

Reply:
I don't have strong feelings on this.  `%f` isn't in the standard anyways (POSIX does not believe in subseconds) so it's a deviation either way.

Reply:
It would be nice to have an option to match Python here. Perhaps we could try to add `%f` to `date.h`.

Reply:
Possible relevant: Related: https://github.com/HowardHinnant/date/issues/562

Reply:
[Looks `date.h` is going a different way than `%f`](https://github.com/HowardHinnant/date/issues/562#issuecomment-608406611). We could mimic that to a degree and add a `precision` parameter to add a possibility to round to `s, ms, us, ns` precisions `(1, 1.123, 1.123456, 1.123456789)`. Unfortunately that only allows decimal place setting in increments of 3.